src/lib.rs

```diff
@@ -463,6 +468,122 @@ impl<'a, T: 'a> NodeMut<'a, T> {
+    where
+        F: FnMut(&mut NodeMut<T>),
+    {
+        if let Some(mut next_sibling_node) = self.next_sibling() {
+            f(&mut next_sibling_node);
+            while let Ok(mut n) = next_sibling_node.into_next_sibling() {
+                f(&mut n);
+                next_sibling_node = n;
+            }
+        }
+    }
+
+    /// Apply function for each previout sibling mutable node reference.
+    pub fn for_each_prev_sibling<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&mut NodeMut<T>),
+    {
+        if let Some(mut prev_sibling_node) = self.prev_sibling() {
+            f(&mut prev_sibling_node);
+            while let Ok(mut n) = prev_sibling_node.into_prev_sibling() {
+                f(&mut n);
+                prev_sibling_node = n;
+            }
+        }
+    }
+
+    /// Apply function for this node and each sibling mutable node reference.
+    pub fn for_each_sibling<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&mut NodeMut<T>),
+    {
+        f(self);
+        self.for_each_prev_sibling(&mut f);
+        self.for_each_next_sibling(&mut f);
+    }
+
+    /// Apply function for each children mutable node reference.
+    pub fn for_each_child<F>(&mut self, f: F)
+    where
+        F: FnMut(&mut NodeMut<T>),
+    {
+        if let Some(mut first_child) = self.first_child() {
+            first_child.for_each_sibling(f);
+        }
+    }
+
+    /// Apply function for this node and each descendant mutable node reference.
+    pub fn for_each_descendant<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&mut NodeMut<T>),
+    {
+        let id = self.id();
+
+        f(self);
+
+        // Start at our first child, if any.
+        if let Some(mut node) = self.first_child() {
+            loop {
+                f(&mut node);
+
+                // Try to go deeper into its first child.
+                match node.into_first_child() {
+                    Ok(child) => {
+                        node = child;
+                        continue;
+                    }
+                    Err(n) => {
+                        node = n;
+                    }
+                }
+
+                // No deeper child, so climb until we find a next sibling or hit self.
+                loop {
+                    match node.into_next_sibling() {
+                        Ok(sib) => {
+                            node = sib;
+                            break;
+                        }
+                        Err(n) => {
+                            node = n;
+                        }
+                    }
+
+                    // No sibling, so climb up.
+                    match node.into_parent() {
+                        Ok(parent) => {
+                            if parent.id() == id {
+                                return;
+                            }
+                            node = parent;
+                        }
+                        Err(_) => {
+                            return;
```

Comment:
Is this really just a return or rather `unrechable!()`?

Reply:
It is, indeed, unreachable! I was also thinking about using `unwrap_unchecked()`, but `unreachable!()` will do as well.

Reply:
Actually, just using `unwrap` on the `into_parent` should suffice.

Reply:
I think it better to keep current approach, since `unwrap` requires `T` to implement `Debug`.

Reply:
Fair.